### PR TITLE
Realtime ASR widget [new]

### PR DESCRIPTION
**Pull request by mishig25**
Creating new PR, rather than continuing on https://github.com/huggingface/huggingface_hub/pull/594 because the structure change of the `huggingface_hub/js` structure.

Apologies for not getting this feature done faster. But excited to finish it this time

# Add `realtime` capability to ASR widget

# How to test locally:
1. hardcode your token [here](https://github.com/huggingface/huggingface_hub/blob/ab27487f9fc9b7ea7f567a4f2e18fbe290479194/js/src/lib/components/InferenceWidget/shared/WidgetRealtimeRecorder/Recorder.ts#L16-L17)
by your token, I mean this value from hf.co/settings/tokens
Screenshot 2022-03-04 at 10 01 43

2. ` cd huggingface_hub/js`
3. `npm ci`
4. `npm run dev`
5. visit http://localhost:3000/facebook/wav2vec2-base-960h

## Screenshots
| initial state | success state |
|---|---|
| Screenshot 2022-01-16 at 17 09 48 | Screenshot 2022-01-16 at 17 09 36 |
| token error | mic access error |
| Screenshot 2022-01-16 at 17 10 04 | Screenshot 2022-01-16 at 17 10 34 |
| warning: mic input is empty | warning: realtime asr output is empty |
| Screenshot 2022-01-16 at 17 11 08 | Screenshot 2022-01-16 at 17 11 32 |

### CODE REVIEW

File: js/static/audioProcessor.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
+class AudioDataExtractor extends AudioWorkletProcessor {
```
Comment (mishig25) [inline js/static/audioProcessor.js:1]: read more [here](https://developer.mozilla.org/en-US/docs/Web/API/AudioWorkletProcessor)

File: js/src/lib/components/InferenceWidget/shared/WidgetRealtimeRecorder/Recorder.ts [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+			throw new Error("You need to be loggedn in and have API token enabled. Find more at: hf.co/settings/token");
+		}
+
+		const constraints: MediaStreamConstraints =
+			this.type === "video"
+				? { audio: true, video: true }
+				: { audio: true };
+		this.stream = await navigator.mediaDevices.getUserMedia(constraints);
+
+		this.socket = new WebSocket(`wss://api-inference.huggingface.co/asr/live/cpu/${this.modelId}`);
```
Comment (mishig25) [inline js/src/lib/components/InferenceWidget/shared/WidgetRealtimeRecorder/Recorder.ts:35]: using `cpu` all the time for now
read more [here](https://github.com/huggingface/huggingface_hub/pull/594#issuecomment-1018937577)

**Comment by mishig25** [comment]
Use this python script from Nicolas if you want to test it locally through python
```py
import datetime
import platform
import subprocess
import os
import base64
import asyncio
import json
import websockets
import curses

# nlp = pipeline("automatic-speech-recognition", device=0)


start = datetime.datetime.now()

sampling_rate = 48_000
chunk_length_s = 0.1

MODEL_ID = "facebook/wav2vec2-large-960h-lv60-self"
COMPUTE_TYPE = "cpu"  # or "gpu"
SENT = 0
AUTH = True
# URI = f"ws://localhost:8000/asr/live/{COMPUTE_TYPE}/{MODEL_ID}"
# URI = f"ws://localhost:8001/asr/live"
URI = f"wss://api-inference.huggingface.co/asr/live/{COMPUTE_TYPE}/{MODEL_ID}"


async def send(websocket):
    global SENT, AUTH
    # You need to login with a first message as headers are not forwarded
    # for websockets
    if AUTH:
        await websocket.send(f"Bearer {os.getenv('HF_API_TOKEN')}")
    format_for_conversion = "f32le"
    microphone = ffmpeg_microphone(
        sampling_rate=sampling_rate, chunk_length_s=chunk_length_s, format_for_conversion=format_for_conversion
    )
    print("> Start talking")
    for i, bytes_audio in enumerate(microphone):
        inputs = {
            "raw": base64.b64encode(bytes_audio).decode("utf-8"),
            "sampling_rate": sampling_rate,
            "format": format_for_conversion,
        }
        # payload = {"id": i, "inputs": inputs}
        await websocket.send(json.dumps(inputs))
        SENT += 1
        await asyncio.sleep(0.01)


async def recv(websocket, stdscr):
    global SENT
    full_text = ""
    while True:
        data = await websocket.recv()
        payload = json.loads(data)
        stdscr.addstr(0, 0, f"Sent {SENT} -> {payload.get('partial', True)}")
        text = payload.get("text", "")
        stdscr.addstr(1, 0, repr(payload))
        stdscr.addstr(2, 0, full_text + text)
        if not payload.get("partial", True):
            full_text += text

        stdscr.clrtoeol()
        stdscr.refresh()


def ffmpeg_microphone(
    sampling_rate: int,
    chunk_length_s: float,
    format_for_conversion: str = "f32le",
):
    ar = f"{sampling_rate}"
    ac = "1"
    if format_for_conversion == "s16le":
        size_of_sample = 2
    elif format_for_conversion == "f32le":
        size_of_sample = 4
    else:
        raise ValueError("Unhandled format `{format_for_conversion}`. Please use `s16le` or `f32le`")

    system = platform.system()
    if system == "Linux":
        format_ = "alsa"
        input_ = "default"
    elif system == "Darwin":
        format_ = "avfoundation"
        input_ = ":0"
    elif system == "Windows":
        format_ = "dshow"
        input_ = "default"

    ffmpeg_command = [
        "ffmpeg",
        "-f",
        format_,
        "-i",
        input_,
        "-ac",
        ac,
        "-ar",
        ar,
        "-f",
        format_for_conversion,
        "-fflags",
        "nobuffer",
        "-hide_banner",
        "-loglevel",
        "quiet",
        "pipe:1",
    ]
    chunk_len = int(round(sampling_rate * chunk_length_s)) * size_of_sample
    iterator = _ffmpeg_stream(ffmpeg_command, chunk_len)
    for item in iterator:
        yield item


def _ffmpeg_stream(ffmpeg_command, buflen: int):
    """
    `bufout` is used to enable streaming partial
    """
    bufsize = 2 ** 24  # 16Mo
    try:
        with subprocess.Popen(ffmpeg_command, stdout=subprocess.PIPE, bufsize=bufsize) as ffmpeg_process:
            while True:
                raw = ffmpeg_process.stdout.read(buflen)
                if raw == b"":
                    break
                yield raw
    except FileNotFoundError as error:
        raise ValueError("ffmpeg was not found but is required to stream audio files from filename") from error


async def main(stdscr):
    curses.noecho()
    curses.cbreak()
    async with websockets.connect(URI) as websocket:
        # inputs and parameters are classic, "id" is a way to track that query
        future = send(websocket)
        future_r = recv(websocket, stdscr)
        await asyncio.gather(future, future_r)


def mmain(stdscr):
    loop = asyncio.get_event_loop()
    if loop.is_running():
        # When running in notebooks
        import nest_asyncio

        nest_asyncio.apply()
    results = loop.run_until_complete(main(stdscr))


curses.wrapper(mmain)
```

**Comment by mishig25** [comment]
Questions:
1. As we decided on [this message](https://huggingface.slack.com/archives/C02G13FEMDH/p1645627112408419?thread_ts=1645554679.974179&cid=C02G13FEMDH) , the pipeline should work with "Bearer" for non-logged in users. Currently, the websocket fails. You can try using the [script](https://github.com/huggingface/huggingface_hub/pull/738#issuecomment-1059018039) you provided above
2. When I change from
```py
MODEL_ID = "facebook/wav2vec2-base-960h"
```
to:
```py
MODEL_ID = "facebook/wav2vec2-large-960h-lv60-self"
```,
the websocket outputs are empty

cc: @Narsil

File: js/src/lib/components/InferenceWidget/shared/WidgetRealtimeRecorder/Recorder.ts (hunk 2 above)
Comment (patrickvonplaten) [reply js/src/lib/components/InferenceWidget/shared/WidgetRealtimeRecorder/Recorder.ts:35]: CPU is good enough for now I think!

File: js/src/lib/interfaces/Types.ts [hunk 3]
```diff
@@ -225,6 +225,7 @@ export interface ModelData {
 		inference?: boolean | {
 			parameters?: Record<string, any>;
 		};
+		widget_realtime_asr?: boolean;
```
Comment (mishig25) [inline js/src/lib/interfaces/Types.ts:228]: The idea is to show realtime widget feature only on models that enabled this flag

cc: @julien-c @osanseviero

File: js/static/audioProcessor.js (hunk 1 above)
Comment (Pierrci) [reply js/static/audioProcessor.js:1]: why a js file here and not a typescript one?

**Comment by Narsil** [comment]
Looking at this tomorrow !

**Comment by Narsil** [comment]
Okay, the reason you were not seeing it is linked to a simple bug that occured since a code change in the API, fix coming up.

Also removal of  token is done, see attached PRs.

**Comment by mishig25** [comment]
Will do so 😊

**Comment by patrickvonplaten** [comment]
Gently pinging - any progress here? :-)

**Comment by Narsil** [comment]
And the main blocker is that I broke some integration test, and I am slowly fixing those.

**Comment by Narsil** [comment]
@mishig25 Everything back online !

**Comment by Pierrci** [comment]
@mishig25 see https://github.com/huggingface/huggingface_hub/pull/738#discussion_r822200677 :)

**Comment by mishig25** [comment]
superceded by https://github.com/huggingface/hub-docs/pull/120

**Pull request closed without merging**